### PR TITLE
Adopting nRF SoC and boards to HWMv2

### PR DESCRIPTION
**Pull request by tejlmand**
Followup on #50305 with porting of Nordic's nRF SoCs and boards.

### CODE REVIEW

**Review by rerickson1 — COMMENTED** [review]
Laird Connectivity boards look good.

I noticed `rm1xx_dvk` is missing from being moved to v2. It is based on `SOC_NRF51822_QFAC`

**Comment by tejlmand** [comment]
compliance and doc build failures are related to `MAINTAINERS.yml` file.

This file will be independently updated when socs and boards have been ported.

Compliance and doc build errors
> get_maintainer.MaintainersError: MAINTAINERS.yml: glob pattern 'boards/arm/bl65*/' in 'files' in area 'Laird Connectivity platforms' does not match any files

**Comment by nordicjm** [comment]
Remaining test failures:

- 96b_carbon needs stm32f4xx porting to hwmv2
- bsim needs soc/boards v2 folder location to be set for the final location so bsim files can be updated

File: boards/v2/nordic_nrf/thingy53/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+* a full-featured Arm Cortex-M33F core with DSP instructions, FPU, and
+  Armv8-M Security Extension, running at up to 128 MHz, referred to as
+  the **application core**
+* a secondary Arm Cortex-M33 core, with a reduced feature set, running at
+  a fixed 64 MHz, referred to as the **network core**.
+
+The ``nrf5340dk/nrf5340/cpuapp`` build target provides support for the application
+core on the nRF5340 SoC. The ``nrf5340dk/nrf5340/cpunet`` build target provides
+support for the network core on the nRF5340 SoC.
```
Comment (kartben) [inline boards/v2/nordic_nrf/thingy53/doc/index.rst:23]: s/nrf5340dk/thingy53/ ?

File: boards/v2/nordic_nrf/thingy52/doc/index.rst
Comment (kartben) [inline boards/v2/nordic_nrf/thingy52/doc/index.rst:1]: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/nrf/battery/src/battery.c#L26-L37 will need updating

**Review by kartben — DISMISSED** [review]

File: boards/v2/96boards/96b_nitrogen/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+Flashing an Application to 96Boards Nitrogen
+============================================
+
+Here is an example for the :ref:`hello_world` application. This
+requires installing the :ref:`pyocd-debug-host-tools`.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: 96b_nitrogen
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply boards/v2/96boards/96b_nitrogen/doc/index.rst]: Some but others no, this one does not

File: boards/v2/nordic_nrf/thingy53/doc/index.rst (hunk 1 above)
Comment (nordicjm) [reply boards/v2/nordic_nrf/thingy53/doc/index.rst]: Fixed

**Comment by kartben** [comment]
Why is the top level folder not "nordic"?
I would think we really want the top-level folder to be something that matches what we have in vendor-prefixes, no? Or to put it differently: what's the expected semantics of `X` in `boards/X` (when X != `shields`, I guess 🙃)

**Comment by nordicjm** [comment]
@kartben please re-review

File: soc/nordic_nrf/nrf51/Kconfig [hunk 3]
```diff
@@ -7,10 +7,7 @@
 config SOC_SERIES_NRF51X
 	select ARM
 	select CPU_CORTEX_M0
-	select SOC_FAMILY_NRF
 	imply XIP
 	select HAS_NRFX
 	select HAS_SEGGER_RTT if ZEPHYR_SEGGER_MODULE
 	select HAS_POWEROFF
-	help
-	  Enable support for NRF51 MCU series
```
Comment (gmarull) [inline soc/nordic_nrf/nrf51/Kconfig]: why does 87c82c111c46bda97607048e8a320b7bdf950897 contain changes to other soc series than 54?

File: soc/nordic_nrf/nrf54h/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Nordic Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_ARM)
+  add_subdirectory(arm)
+elseif(CONFIG_RISCV)
+  add_subdirectory(riscv)
+endif()
```
Comment (gmarull) [inline soc/nordic_nrf/nrf54h/CMakeLists.txt:8]: why is this needed?

File: soc/nordic_nrf/nrf51/Kconfig (hunk 3 above)
Comment (nordicjm) [reply soc/nordic_nrf/nrf51/Kconfig]: Because the original port for nordic was not done properly and this was needed to actually port it properly

File: soc/nordic_nrf/nrf54h/CMakeLists.txt (hunk 4 above)
Comment (nordicjm) [reply soc/nordic_nrf/nrf54h/CMakeLists.txt:8]: to include the cmake file for the correct core

File: soc/nordic_nrf/nrf54h/arm/Kconfig.soc [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Nordic Semiconductor nRF54H MCU line
+
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_NRF54HX
+	bool
+	select SOC_FAMILY_NRF
+	help
+	  Nordic Semiconductor nRF54H series MCU
```
Comment (gmarull) [inline soc/nordic_nrf/nrf54h/arm/Kconfig.soc:10]: if we preserve the arch folder split under soc/nordic_nrf, we'll need to keep duplicating stuff like that, which to me, seems unnecessary. IMHO, we should have all CPUs defined in a single `Kconfig.soc`, regardless of their arch.

File: soc/nordic_nrf/nrf54h/riscv/Kconfig.soc [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_NRF54HX
+	bool
+	select SOC_FAMILY_NRF
+	help
+	  Nordic Semiconductor nRF54H series MCU
```
Comment (gmarull) [inline soc/nordic_nrf/nrf54h/riscv/Kconfig.soc:8]: see my comment above

File: soc/nordic_nrf/nrf54h/arm/Kconfig [hunk 7]
```diff
@@ -3,32 +3,29 @@
...
-
-choice
-	prompt "nRF54H20 MCU Selection"
+config SOC_SERIES_NRF54HX
+	select ARM
+	select ARMV8_M_DSP
+	select CPU_CORTEX_M33
+	select HAS_NRFX
+	select HAS_NORDIC_DRIVERS
+	select HAS_SEGGER_RTT if ZEPHYR_SEGGER_MODULE
```
Comment (gmarull) [inline soc/nordic_nrf/nrf54h/arm/Kconfig:12]: I find this confusing, because the NRF54HX SoC series no longer can select ARM (or CPU_CORTEX_M33) in a multi-arch setup. These options should be selected by each CPU, even if it implies duplication (imagine eg we had a CORTEX-M33, a CORTEX-M0 and a VPR RISCV).

File: soc/nordic_nrf/nrf51/Kconfig (hunk 3 above)
Comment (gmarull) [reply soc/nordic_nrf/nrf51/Kconfig]: then please, split commits, amend original commits, change message or anything else that makes things clear.

Comment (nordicjm) [reply soc/nordic_nrf/nrf51/Kconfig]: Commit message updated

File: soc/nordic_nrf/nrf54h/arm/Kconfig.soc (hunk 5 above)
Comment (nordicjm) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: Have moved the Kconfig files to the nrf54h directory, the directory split is still needed for arm and risc-specific files e.g. soc.c

File: soc/nordic_nrf/nrf54h/arm/Kconfig (hunk 7 above)
Comment (nordicjm) [reply soc/nordic_nrf/nrf54h/arm/Kconfig]: Split it up properly (this would be easy to spot if there was a board supporting cpuppr)

File: soc/nordic_nrf/nrf54h/arm/Kconfig.soc (hunk 5 above)
Comment (gmarull) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: why don't we simply have `soc_arm.c` and `soc_riscv.c`? As said, the directory split and implied Kconfig duplication just feels wrong. Moving to a hwmv2 where there are arch folders within vendor soc folder isn't a great model.

Comment (nordicjm) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: What file would `#include <soc.h>` which is used all over the place include?

Comment (gmarull) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: Nordic RISC-V code doesn't have `soc.h` since it's fortunately exempt from the [CMSIS mess](https://github.com/zephyrproject-rtos/zephyr/issues/57246). In fact, nobody should be using `soc.h` now that there's `cmsis_core.h` including it. The whole soc.h idea is so brain-damaging.

Comment (gmarull) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: they should simply have a single soc.h, it's now ignored by RISC-V arch code see e7cc2fafb49c1b760bdd85ebf2bc67c53b30c031. If they have unnecessary stuff in soc.h, they should just put it somewhere else. soc.h should just exist to include the "CMSIS glue code" (the entire HAL in most cases)

Comment (gmarull) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: Btw, the whole soc.h scheme is so messed up that stuff like https://github.com/zephyrproject-rtos/zephyr/commit/00566d64fa901cbd7725068f6ff82ee0f0fffc16 is needed because soc.h pollutes almost every single compilation unit.

Comment (nordicjm) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: then you would still have all these files: https://github.com/tejlmand/zephyr/tree/nrf52_adopt_soc_scheme_v2/soc/nordic_nrf/nrf54h/riscv/common/vpr

Renaming them all to `riscv_<filename>` would look like a mess

Comment (gmarull) [reply soc/nordic_nrf/nrf54h/arm/Kconfig.soc]: These files should be under `soc/nordic_nrf/common/vpr`, since the VPR core is present in eg nRF54L15 (not added yet). In the future, we should think about supporting RISC-V core setups under `arch/riscv/cores` or something similar, we could have scenarios where a certain RISC-V core implementation is supported by multiple different vendor SoCs.

**Review by gmarull — APPROVED** [review]
ack for nrf54x changes, a post-hwmv2 cleanup can likely improve soc/nordic_nrf a bit

**Review by carlescufi — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

File: cmake/modules/configuration_files.cmake [hunk 8]
```diff
@@ -77,8 +77,20 @@ zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS APP_BOARD_DTS SUFFIX
...
+
+  if(DEFINED BOARD_IDENTIFIER)
+    # Generate overlay filename of board with identifier present
+    zephyr_build_string(board_with_identifier_overlay
+                        BOARD ${BOARD}
+                        BOARD_IDENTIFIER ${BOARD_IDENTIFIER}
+    )
+
+    set(board_with_identifier_overlay "${board_with_identifier_overlay}.overlay")
+  endif()
```
Comment (tejlmand) [inline cmake/modules/configuration_files.cmake:90]: How about:
```suggestion
  zephyr_build_string(board_overlay_strings
                      BOARD ${BOARD}
                      BOARD_IDENTIFIER ${BOARD_IDENTIFIER}
                      MERGE
  )
  list(TRANSFORM board_overlay_strings APPEND ".overlay")
```
and then use `board_overlay_strings` below in:

```
        NAMES "${APP_BOARD_DTS};${board_overlay_strings };app.overlay" SUFFIX ${FILE_SUFFIX})
```

Comment (nordicjm) [reply cmake/modules/configuration_files.cmake]: Fixed
